File: src/Moka.Red.Primitives/Utility/MokaScrollToTop.razor.js
/**
 * MokaScrollToTop JS module - watches a scroll position and scrolls back to the top.
 *
 * The position is the window's, or that of the element a CSS selector names, for layouts where
 * the page content scrolls inside a panel. Every init() gets its own handle and listener, so two
 * buttons on one page do not tear each other down. Callers pass the handle back to scrollToTop()
 * and dispose().
 */

const watchers = new Map();
let nextHandle = 1;

/**
 * Starts watching a scroll position for one component instance.
 * @param {object} dotNetRef - .NET object reference for callbacks.
 * @param {number} showAfter - Scroll offset in pixels past which the button shows.
 * @param {string|null} selector - CSS selector of the element that scrolls, or null for the window.
 * @returns {number} Handle to pass to scrollToTop() and dispose().
 */
export function init(dotNetRef, showAfter, selector) {
    const handle = nextHandle++;
    const watcher = {
        dotNetRef,
        showAfter,
        selector: selector || null,
        element: null,
        visible: null,
        onScroll: null
    };
    watchers.set(handle, watcher);

    if (!watcher.selector) {
        watcher.onScroll = () => report(watcher, window.scrollY);
        window.addEventListener('scroll', watcher.onScroll, { passive: true });
        report(watcher, window.scrollY);
        return handle;
    }

    if (!isValidSelector(watcher.selector)) {
        console.warn(`MokaScrollToTop: "${watcher.selector}" is not a valid CSS selector, so the button never shows.`);
        report(watcher, Number.NEGATIVE_INFINITY);
        return handle;
    }

    // Scroll events do not bubble, but a capturing listener on the document still receives them.
    // Matching the target on each event keeps working after a re-render replaces the element.
    watcher.onScroll = event => {
        const target = event.target;
        if (target instanceof Element && target.matches(watcher.selector)) {
            watcher.element = target;
            report(watcher, target.scrollTop);
        }
    };
    document.addEventListener('scroll', watcher.onScroll, { capture: true, passive: true });

    // Without the element there is nothing to scroll, so the button stays hidden until it scrolls.
    const element = document.querySelector(watcher.selector);
    watcher.element = element;
    report(watcher, element ? element.scrollTop : Number.NEGATIVE_INFINITY);
    return handle;
}

/**
 * Scrolls the watched element, or the window, back to the top.
 * @param {number} handle - The value init() returned.
 * @param {boolean} smooth - Whether to animate the scroll.
 */
export function scrollToTop(handle, smooth) {
    const watcher = watchers.get(handle);
    const target = watcher ? targetOf(watcher) : window;
    target?.scrollTo({ top: 0, behavior: smooth ? 'smooth' : 'instant' });
}

/**
 * Removes the listener registered by the matching init() call.
 * @param {number} handle - The value init() returned.
 */
export function dispose(handle) {
    const watcher = watchers.get(handle);
    if (!watcher) return;

    if (watcher.onScroll) {
        if (watcher.selector) {
            document.removeEventListener('scroll', watcher.onScroll, { capture: true });
        } else {
            window.removeEventListener('scroll', watcher.onScroll);
        }
    }

    watchers.delete(handle);
}

// Calls .NET only when the button should appear or disappear, not on every scroll event, which
// on Blazor Server would be a message per frame while scrolling.
function report(watcher, offset) {
    const visible = offset > watcher.showAfter;
    if (visible === watcher.visible) return;

    watcher.visible = visible;
    watcher.dotNetRef.invokeMethodAsync('OnScrollChanged', visible);
}

function targetOf(watcher) {
    if (!watcher.selector) return window;
    if (watcher.element?.isConnected) return watcher.element;
    return isValidSelector(watcher.selector) ? document.querySelector(watcher.selector) : null;
}

function isValidSelector(selector) {
    try {
        document.createDocumentFragment().querySelector(selector);
        return true;
    } catch {
        return false;
    }
}

File: src/Moka.Red.Primitives/Utility/MokaScrollToTop.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using Moka.Red.Core.Utilities;

namespace Moka.Red.Primitives.Utility;

/// <summary>
///     Fixed-position button that appears once the window, or the element named by
///     <see cref="ScrollContainerSelector" />, has scrolled down, and scrolls it back to the top on click.
/// </summary>
public partial class MokaScrollToTop
{
	private const string ModulePath = "./_content/Moka.Red.Primitives/Utility/MokaScrollToTop.razor.js";

	private bool _disposed;
	private DotNetObjectReference<MokaScrollToTop>? _dotNetRef;
	private int _scrollHandle;
	private bool _visible;

	// What the JS listener was set up with. A change to either parameter moves the listener.
	private (int ShowAfter, string? Selector)? _watching;
	private Task _watchSync = Task.CompletedTask;

	/// <summary>Pixels scrolled before showing the button. Default 200.</summary>
	[Parameter]
	public int ShowAfter { get; set; } = 200;

	/// <summary>Use smooth scroll animation. Default true.</summary>
	[Parameter]
	public bool Smooth { get; set; } = true;

	/// <summary>
	///     CSS selector of the element that scrolls, such as <c>".app-main"</c>. The button then
	///     watches and scrolls that element instead of the window. Set it when the page content
	///     scrolls inside a panel, as in dock layouts and app shells, where the window never scrolls.
	///     Default null, the window.
	/// </summary>
	[Parameter]
	public string? ScrollContainerSelector { get; set; }

	/// <summary>Tooltip and accessible name of the button. Default "Scroll to top".</summary>
	[Parameter]
	public string Label { get; set; } = "Scroll to top";

	/// <inheritdoc />
	protected override string RootClass => "moka-scroll-top";

	/// <inheritdoc />
	protected override string CssClass => new CssBuilder(RootClass)
		.AddClass(Class)
		.Build();

	/// <inheritdoc />
	protected override Task OnAfterRenderAsync(bool firstRender)
	{
		base.OnAfterRender(firstRender);

		// One sync at a time: the first report from JS renders again before init has returned, and
		// that render must not start a second listener.
		_watchSync = WatchAsync(_watchSync);
		return _watchSync;
	}

	/// <summary>Called from JS when the button should appear or disappear.</summary>
	[JSInvokable]
	public void OnScrollChanged(bool visible)
	{
		if (_visible != visible)
		{
			_visible = visible;
			InvokeAsync(ForceRender);
		}
	}

	private async Task WatchAsync(Task previous)
	{
		await previous;

		(int ShowAfter, string? Selector) wanted = (ShowAfter, ScrollContainerSelector);
		if (_disposed || _watching == wanted)
		{
			return;
		}

		await StopWatchingAsync();
		_dotNetRef ??= DotNetObjectReference.Create(this);
		_scrollHandle = await SafeModuleInvokeAsync<int>(ModulePath, "init", _dotNetRef, ShowAfter,
			ScrollContainerSelector);
		_watching = wanted;
	}

	private async Task StopWatchingAsync()
	{
		if (_scrollHandle != 0)
		{
			await SafeModuleInvokeVoidAsync(ModulePath, "dispose", _scrollHandle);
			_scrollHandle = 0;
		}

		_watching = null;
	}

	private async Task ScrollToTop()
	{
		// The disabled attribute stops clicks in the browser; this also covers events raised another way.
		if (Disabled)
		{
			return;
		}

		await SafeModuleInvokeVoidAsync(ModulePath, "scrollToTop", _scrollHandle, Smooth);
	}

	/// <inheritdoc />
	protected override async ValueTask DisposeAsyncCore()
	{
		// Let an attach in flight finish, then drop the scroll listener before the base disposes the
		// module. Otherwise it stays attached for the life of the page, calling a dead component.
		_disposed = true;
		try
		{
			await _watchSync;
		}
		catch (JSException)
		{
			// The attach failed in the browser, so there is no listener to remove.
		}

		await StopWatchingAsync();

		_dotNetRef?.Dispose();
		await base.DisposeAsyncCore();
	}
}
